lib/src/rules/conformance/web_bindings.dart

```diff
@@ -0,0 +1,188 @@
+/// those. If the class has a `@Native` annotation and does not have any values,
+/// it returns a list with just the Dart name. Otherwise, returns null.
+List<String>? _getNativeTypes(ClassElement element) {
+  for (var annotation in element.metadata) {
+    var annotationElement = annotation.element;
+    if (annotationElement is ConstructorElement &&
+        annotationElement.enclosingElement.name == _nativeAnnotationName &&
+        annotationElement.library.name == _dartJsHelperLibrary) {
+      var value =
+          annotation.computeConstantValue()?.getField('name')?.toStringValue();
+      if (value != null && value.isNotEmpty) {
+        return value.split(',');
+      } else {
+        return [element.name];
+      }
+    }
+  }
+  return null;
+}
+
+/// Returns the value in the `@JSName` annotation.
+///
+/// If the annotation exists and has a value, returns that value. Otherwise,
+/// returns null.
+String? _getJsNameValue(List<ElementAnnotation> metadata) {
+  for (var annotation in metadata) {
+    var annotationElement = annotation.element;
+    if (annotationElement is ConstructorElement &&
+        annotationElement.enclosingElement.name == _jsNameAnnotationName &&
+        annotationElement.library.name == _dartJsHelperLibrary) {
+      var value =
+          annotation.computeConstantValue()?.getField('name')?.toStringValue();
+      if (value != null && value.isNotEmpty) {
+        return value;
+      }
+    }
+  }
+  return null;
+}
+
+// TODO(srujzs): The following maps are roughly a couple MB (depending on
+// encoding), as there are hundreds of types bound in `dart:html` with a varying
+// number of properties per type. If this is too large, an alternative would be
+// to do a lookup in the AST every time, but this adds considerable computation
+// time for every rule.
+
+// Mapping of native types that are bound in `dart:html` to a map of their
+// properties to those properties' Dart names.
+final Map<String, Map<String, String>> _nativeTypeToDartProperties = {};
+// Mapping of types in `dart:html` to a map of their properties to the native
+// properties that those properties bind to.
+final Map<String, Map<String, String>> _dartTypeToNativeProperties = {};
+// Mapping of types in `dart:html` to the native types they bind.
+final Map<String, List<String>> _dartTypeToNativeTypes = {};
+
+/// Computes and caches native property bindings from `dart:html`.
+///
+/// Specifically, computes the three maps above. Given the current [library],
+/// uses its imports and exports to find the web libraries. This mapping is
+/// necessary if the library dynamically calls a setter, since the property name
+/// may match a renamed property in `dart:html`, and we would have no reference
+/// to that binding without this processing. It's also necessary to determine
+/// whether a native type could be used with non-static interop.
+void computeHtmlBindings(LibraryElement? library) {
```

Comment:
Really it boils down to whether we want to support the two clauses in the comments. The latter case (`whether a native type could be used with non-static interop`) is much less common today, but it might be increasingly common when we refactor things out of `dart:html`. It also needs way less information stored, since it only cares about the list of `@Native` types and not any properties.

The former case is the sticking point, unfortunately. It needs both types and properties to be accurate. It's possible if it's a big enough issue, we can discuss if we're okay ignoring these two instances, but we can talk about that more on Monday.

I do like the idea of lazily computing the cache. Let me try moving all these to a class and populating when needed and respond back.

Reply:
Moved this processing to a static instance of a class that lazily computes these bindings when needed. I expect this static instance to move to shared code in the future, but for now it can live within the visitor.